File: server/middleware/security.js
'use strict';

const helmet = require('helmet');

/*
 * Restrictive CSP: no external CDNs (all libs, including pdf-lib and
 * fflate, are served locally from /js/vendor). No inline scripts/styles
 * beyond what's needed, no third-party frames.
 */
const securityHeaders = helmet({
  contentSecurityPolicy: {
    directives: {
      defaultSrc: ["'self'"],
      scriptSrc: ["'self'"],
      styleSrc: ["'self'", "'unsafe-inline'"], // minimal inline styles in the static HTML
      imgSrc: ["'self'", 'blob:', 'data:'],
      mediaSrc: ["'self'", 'blob:'],
      connectSrc: ["'self'", 'ws:', 'wss:', 'https://api.github.com'], // landing downloads: reads releases/latest
      fontSrc: ["'self'"],
      objectSrc: ["'none'"],
      baseUri: ["'none'"],
      formAction: ["'self'"],
      frameAncestors: ["'none'"],
      workerSrc: ["'self'", 'blob:'],
      upgradeInsecureRequests: [],
    },
  },
  referrerPolicy: { policy: 'no-referrer' },
  frameguard: { action: 'deny' },
  crossOriginEmbedderPolicy: false,
  crossOriginResourcePolicy: { policy: 'same-origin' },
});

module.exports = { securityHeaders };

File: standalone/security.go
package main

import "net/http"

// Restrictive CSP — same directives as server/middleware/security.js
// (no external CDNs: the vendor assets don't exist anymore, everything
// left is local). Plus the hardening headers helmet applied by default on
// the Node side.
const csp = "default-src 'self'; " +
	"script-src 'self'; " +
	"style-src 'self' 'unsafe-inline'; " +
	"img-src 'self' blob: data:; " +
	"media-src 'self' blob:; " +
	"connect-src 'self' ws: wss: https://api.github.com; " + // landing downloads: reads releases/latest
	"font-src 'self'; " +
	"object-src 'none'; " +
	"base-uri 'none'; " +
	"form-action 'self'; " +
	"frame-ancestors 'none'; " +
	"worker-src 'self' blob:; " +
	"upgrade-insecure-requests"

func securityHeaders(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		h := w.Header()
		h.Set("Content-Security-Policy", csp)
		h.Set("Referrer-Policy", "no-referrer")
		h.Set("X-Frame-Options", "DENY")
		h.Set("X-Content-Type-Options", "nosniff")
		h.Set("Cross-Origin-Resource-Policy", "same-origin")
		h.Set("X-XSS-Protection", "0")
		h.Set("X-Permitted-Cross-Domain-Policies", "none")
		h.Set("Strict-Transport-Security", "max-age=15552000; includeSubDomains")
		next(w, r)
	}
}
